orbit/pkg/table/fleetd_logs/fleetd_logs.go

```diff
@@ -0,0 +1,166 @@
+package fleetd_logs
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/osquery/osquery-go/plugin/table"
+	"github.com/rs/zerolog"
+)
+
+// No timezone, always return in UTC. Use this format because SQLite3
+// knows how to parse it.
+// See https://www.sqlite.org/lang_datefunc.html
+const timeFormatString = "2006-01-02 15:04:05.999999999"
+
+var Logger = logger{}
+var MaxEntries uint = 10_000
+
+func TablePlugin() *table.Plugin {
+	columns := []table.ColumnDefinition{
+		table.TextColumn("time"),
+		table.TextColumn("level"),
+		table.TextColumn("payload"),
+		table.TextColumn("message"),
+		table.TextColumn("error"),
+	}
+
+	return table.NewPlugin("fleetd_logs", columns, generate)
+}
+
+func generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	output := []map[string]string{}
+
+	for _, entry := range Logger.logs {
+		row := make(map[string]string, 5)
+		// It would be nice if we could return NULL instead of an
+		// empty string when the error is empty
+		row["time"] = entry.Time
+		row["level"] = entry.Level.String()
+		row["payload"] = string(entry.Payload)
+		row["message"] = entry.Message
+		row["error"] = entry.Error
+		output = append(output, row)
+	}
+	return output, nil
+}
+
+type Event struct {
+	Time    string
+	Level   zerolog.Level
+	Payload []byte
+	Message string
+	Error   string
+}
+
+type logger struct {
+	writeMutex sync.Mutex
+	logs       []Event
+}
+
+func (l *logger) Write(event []byte) (int, error) {
+	msg, err := processLogEntry(event)
+	if err != nil {
+		return 0, fmt.Errorf("fleet_logs.Write: %w", err)
+	}
+
+	l.writeMutex.Lock()
+	defer l.writeMutex.Unlock()
+
+	l.logs = append(l.logs, msg)
+
+	if MaxEntries > 0 && len(l.logs) > int(MaxEntries) {
+		l.logs = l.logs[len(l.logs)-int(MaxEntries):]
+	}
```

Comment:
can you sanity check my logic here?

1. Once we reached `MaxEntries`, won't this keep allocating and copying `10_000` elements to new arrays for the slices to point to?
2. Minor: can we initialize the slice capacity to avoid reallocation while the slice is filling in for the first time?

Reply:
If you run this you'll see that that capacity keeps shrinking as we shift the array, and it only re-allocates the array once the len and cap match

```go
package main

import (
	"fmt"
	"unsafe"
)

func main() {
	max := 25
	arr := []bool{}
	for range 1000 {
		arr = append(arr, true)
		if len(arr) > max {
			arr = arr[len(arr)-max:]
		}
		fmt.Printf("ptr=%p len=%d, cap=%d\n", unsafe.SliceData(arr), len(arr), cap(arr))
	}
}
```

Reply:
That being said, a ring buffer would probably be a better data structure for this since then we would never need to re-allocate anything

Reply:
yes, that's what I'm worried about (I don't know of a quick and easy way to sanity check). I'm concerned about the underlying array will keep on growing (using amortized constant time, so the copies are not that bad) but I think all the logs will be carried over.

check out this section: https://go.dev/blog/slices-intro#a-possible-gotcha

Reply:
I commented that ^ without seeing your code example, taking a look now

Reply:
If we add 1 million entries to a 10,000 big array, it only reallocs less than 300 times

```go
package main

import (
	"fmt"
	"unsafe"
)

func main() {
	max := 10_000
	arr := []bool{}
	reallocs := 0
	for range 1_000_000 {
		origCap := cap(arr)
		arr = append(arr, true)
		if len(arr) > max {
			arr = arr[len(arr)-max:]
		}
		newCap := cap(arr)
		if origCap < newCap {
			println("reallocated")
			reallocs++
		}
		fmt.Printf("ptr=%p len=%d, cap=%d\n", unsafe.SliceData(arr), len(arr), cap(arr))
	}
	println("Reallocs:", reallocs)
}
```
```
Reallocs: 293
```

Reply:
Still more than we want though perhaps

Reply:
I'm okay with reallocations as they're amortized, I was worried about us keeping _all_ the logs in memory. I think we proved that we're good! thank you!

just for kicks I also did this contrived example: https://go.dev/play/p/0qzpJ54-NR_j